[CBRD-25865] refactor rewriter

sohee-dgist opened the pull request:
http://jira.cubrid.org/browse/CBRD-25865

refactor query rewriter

query_rewriter  ->
- query_rewrite_predication.c
 - query_rewrite_auto_parameterize.c
 - query_rewrite_select.c
 - query_rewrite_set.c
 - query_rewrite_subquery.c
 - query_rewrite_util.h
 - query_rewrite.h
 
To track differences, the following approach was adopted.
The current copied version is based on commit hash 9d6cdce.

Commit:
(refactor) moved query_rewrite.c -> rewriter/query_rewrite.c

Commit:
(refactor) added copy files to change
- rewriter - query_rewrite_predication.c
           - query_rewrite_auto_parameterize.c
           - query_rewrite_select.c
           - query_rewrite_set.c
           - query_rewrite_subquery.c
           - query_rewrite_util.h
           - query_rewrite.h
- To track differences, the following approach was adopted.
- The current copied version is based on commit hash 9d6cdce.

Commit:
(refactor) separate functions to files

Commit:
(refactor) separate logics to function and made buildable

- qo_rewrite_union_with_limit_clause
- qo_rewrite_predications
- qo_rewrite_select_queries

seperated to new functions
and made as buildable

Commit:
(refactor) cleaned header

- include 정리
- 주석 이름 오류 제거

Commit:
(refactor) 주석 오류 해결

- 주석 오류 해결

Commit:
(refactor) gnu coding style, license에 맞추어 수정

- 빈 line 제거
- license앞 공백 제거

Commit:
(refactor) rename predications -> predicate

- do not use predication

Commit:
(refactor) refactoring 오류 해결

refactoring 중 발생한 오류 해결

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

Commit:
(refactor) refactoring 함수화 중 생긴 버그 해결

함수 포인터 관련 수정사항 반영

sohee-dgist commented:
아까 merge 안하고 넣다가 많이 issue가 발생해서
merge 하여 다시 PR 넣었습니다.
실수로 rebase를 사용하는 바람에 push가 불가능한 상태가 되어서요....

Commit:
(refactor) test 오류 해결

- refactoring중 발생한 test 오류 해결

xmilex-git commented on cs/CMakeLists.txt:
```diff
@@ -237,7 +237,13 @@ set(OPTIMIZER_SOURCES
   ${OPTIMIZER_DIR}/query_planner.c
   ${OPTIMIZER_DIR}/plan_generation.c
-  ${OPTIMIZER_DIR}/query_rewrite.c
+  ${OPTIMIZER_DIR}/rewriter/query_rewrite_util.c
```
최상위 CMakeLists.txt에 REWRITER_DIR을 추가해서 OPTIMIZER_DIR과 따로 관리하는건 어떨까요?

sohee-dgist replied:
해당 부분 반영해서 커밋하도록 하겠습니다.

xmilex-git commented on src/optimizer/rewriter/query_rewrite_backup.c:
```diff
@@ -0,0 +1,175 @@
+ */
+
+/*
+ * query_rewrite_backup.c
```
query_rewrite_backup.c 보다는 query_rewrite_unused_functions.c 등으로 관리하는게 알아보기 편할 것 같습니다. 백업 기능과 혼동될 여지가 있어서요..

sohee-dgist replied:
backup 기능과 확실히 혼동될 수 있겠네요, 해당 부분 반영하도록 하겠습니다.

youngjinj replied:
backup 보다 unused는 어떨까요?

xmilex-git commented on src/optimizer/rewriter/query_rewrite_predicate.c:
```diff
@@ -0,0 +1,4114 @@
+ */
+
+/*
+ * query_rewrite_predicate.c
```
코드에 pred, predicate 보다 term으로 표기된 부분이 많아서 term 쪽으로 용어를 변경하는게 혼동을 피하기 좋을것 같습니다. QO(query optimizer) 소스 쪽에서는 term으로 표현하는 경우가 더 많은것 같습니다.

sohee-dgist replied:
확실히 LIKE Term과 같이 Predication과 분리해서 보아야 하는 부분이 조금 존재하네요, 해당 부분 반영하도록 하겠습니다.

shparkcubrid commented:
단순히 분리된 소스코드 말고 수정된 부분은 커멘트해주면 리뷰하는데 도움이 될 것 같네요.

sohee-dgist commented:
최종적으로 수정된 부분.
메인 함수 (qo_optimize_queries-> qo_rewrite_queries)에서

1. query_rewrite중 predication부분을 별도로 분리함.
2. query_rewrite중 select관련 쿼리를 최적화하는 부분을 별도로 분리함.

입니다.
오류를 유발하는 부분은 제외하였습니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_auto_parameterize.c:
```diff
@@ -0,0 +1,369 @@
+
+#ident "$Id$"
+
+#include <assert.h>
+#include "parser.h"
+#include "parser_message.h"
+#include "parse_tree.h"
+#include "optimizer.h"
+
+
+#include "object_primitive.h"
+#include "object_representation.h"
+
+#include "dbtype.h"
```
parser.h 헤더 파일을 제외하고는 필요하지 않은 헤더 파일로 보입니다.
1. 확장성을 고려해서 범용적으로 사용되는 헤더 파일들의 include를 유지하는게 나을까요?
2. 아니면 당장은 parser.h 헤더 파일만 include 가는게 나을까요?

1번처럼 범용적으로 include 되어야 하는 헤더 파일이라고 판단된다면,
각각의 *.c 파일에서 include 하기 보다는 src/optimizer/rewriter/query_rewrite.h 헤더 파일에서 include 하고
#include "query_rewrite.h" 를 include 하는게 좋을 것 같다는 생각입니다.

리뷰하시는 다른 분들의 의견도 궁금합니다.

sohee-dgist replied:
개별로 include를 하면 관리가 어려울 것 같습니다.
영진님 의견인 

각각의 *.c 파일에서 include 하기 보다는 src/optimizer/rewriter/query_rewrite.h 헤더 파일에서 include 하고
#include "query_rewrite.h" 를 include 하는게 좋을 것 같다는 생각입니다.

를 수용하는 방안으로 고치는 것이 좋아 보입니다.

youngjinj replied:
그렇다면 rewriter 쪽에는 항상 include 되어 있어야 하는 범용적인 헤더 파일로 보시나요?
별도로 논의 해보는 것도 좋을 것 같습니다.

shparkcubrid replied:
필요한 헤더파일만 작성하는 게 맞겠죠. 개별 혹은 통합 어떻게 해도 저는 상관없습니다.

youngjinj replied:
확인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_predicate.c:
```diff
@@ -0,0 +1,4114 @@
+ *   expr(in):
+ */
+static int
+qo_is_cast_attr (PT_NODE * expr)
```
사용되지 않는 함수로 보입니다.
`ENABLE_UNUSED_FUNCTION`으로 지정하는 것은 어떨까요?

sohee-dgist replied:
qo_is_cast_attr 은 사용중인 함수이긴 합니다.
qo_reduce_equality_terms내에서 사용중입니다.

youngjinj replied:
아.. 착각했습니다. 확인해주셔서 감사합니다.

Commit:
(refactor) code review 1차 반영

- REWRITER DIR 추가
- query_rewrite_backup -> query_rewrite_unused_function
- query_rewrite_predicate -> query_rewrite_term
- header 정리

youngjinj commented on src/optimizer/rewriter/query_rewrite_util.h:
```diff
@@ -0,0 +1,71 @@
+                                (type) == PT_JOIN_RIGHT_OUTER)
+
+
+#define PT_IS_EXPR_NODE_WITH_COMP_OP(n) \
+        ( (PT_IS_EXPR_NODE (n)) && \
+          ((n)->info.expr.op == PT_EQ || \
+           (n)->info.expr.op == PT_GE || \
+           (n)->info.expr.op == PT_GT || \
+           (n)->info.expr.op == PT_LT || \
+           (n)->info.expr.op == PT_LE || \
+           (n)->info.expr.op == PT_GT_INF || \
+           (n)->info.expr.op == PT_LT_INF || \
+           (n)->info.expr.op == PT_RANGE ))
```
이 매크로는 중복 정의되어 있습니다.
1. src/parser/parse_tree.h
2. src/optimizer/rewriter/query_rewrite_util.h
3. src/optimizer/rewriter/query_rewrite_util.c

parse_tree.h 헤더 파일만 정의하도록 하거나
rewrite와 관련되어 있는 매크로들은 query_rewrite_util.h 헤더 파일에 정의하도록 하는게 좋을 것 같습니다.
다른 분들의 의견도 같이 들어서 절충해주세요.

sohee-dgist replied:
확인했습니다 반영하도록 하겠습니다.

저는 parse_tree.h에만 있어야 한다고 생각합니다.
기존에 왜 존재했었는지를 모르겠네요.

youngjinj replied:
확인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_util.h:
```diff
@@ -0,0 +1,71 @@
+		}                                                \
+	} while (0)
+
+#define  IS_CONDITIONAL_JOIN(type) ((type) == PT_JOIN_INNER || \
```
혹시 IS_CONDITIONAL_JOIN 매크로 함수는 어디에 사용되나요?
AS-IS 코드에서도 찾을 수가 없어서요.

sohee-dgist replied:
제가 리팩토링 하면서 추가하다가 삭제했는데, 누락한 것 같습니다.
의도가 CONDITIONAL JOIN 같아서 해당 매크로로 대체했는데 모종의 이유로 수정하게 되었네요.

제거하는 방향으로 진행하겠습니다.

youngjinj replied:
확인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_util.h:
```diff
@@ -0,0 +1,71 @@
+#ifndef _QUERY_REWRITER_UTIL_H_
+#define _QUERY_REWRITER_UTIL_H_
+
+#define QO_CHECK_AND_REDUCE_EQUALITY_TERMS(parser, node, where) \
```
`query_rewrite_util.h`, `query_rewrite_util.c` 파일에 포함되는 함수들의 기준을 알려주시면 좋을 것 같습니다.

`QO_CHECK_AND_REDUCE_EQUALITY_TERMS` 같은 매크로는
`query_rewrite_predicate.c` 에 포함되면 좋을 것 같은데
`query_rewrite_util.c` 에 포함되어 있습니다.

`qo_is_cast_attr` 함수는
`query_rewrite_util.c` 에 포함되면 좋을 것 같은데
`query_rewrite_predicate.c` 에 포함되어 있습니다.

sohee-dgist replied:
우선, 헤더를 깔끔하게 정리하기 위해 거대한 함수인 qo_rewrite_select_queries와 qo_rewrite_terms 내부에 static(내부 전용)으로 정의되어 있던 모든 내용을 해당 함수가 속한 파일 내로 통합했습니다.

또한, query_rewrite_util.h와 rewrite_util.c 파일은 query_rewriter 모듈에서 여러 파일에 공통적으로 포함되어야 하는 기능 중 특정 기능에 치우치지 않은 일반적인 내용들을 기준으로 이동했습니다.

한편, rewriter util에 포함된 MACRO의 경우는 별도의 파일로 옮기는 것이 적절하다고 판단되어 이를 이동할 예정입니다. 그 외에도 고유한 기능을 담당하는 함수들은 별도로 분리하는 방안을 고려하고 있습니다.
util이 사실 애매해서 util을 삭제하는 방향으로 진행하고자 합니다.

youngjinj replied:
util을 없애는 방향에 동의합니다.
util에 포함되어야 할 함수들이 있더라도 util 보다는
parse_tree 또는 parser_support에 포함되는게 좋지 않을까 하는 생각입니다.

변경된 후에 알려주시면 다시 리뷰를 시작하겠습니다.

sohee-dgist replied:
qo_is_oid_const
qo_rewrite_hidden_col_as_derived
qo_rewrite_index_hints

가 애매하여, 해당 세 함수를 제외하고 모두 이동시켰습니다.

util.h의 삭제 작업을 추가하고 중복되는 매크로는 제거 완료하였습니다.

youngjinj replied:
`qo_is_oid_const`
`qo_rewrite_hidden_col_as_derived`

이 2개는 `src/optimizer/rewriter/query_rewrite_select.c` 로 옮기는게 좋을 것 같습니다.

`qo_rewrite_index_hints` 애매하네요.

UPDATE, DELETE 문이라고 해도 내부적으로는 SELECT 문으로 변경해서 실행합니다.
SELECT 문에 적용하기 위한 인덱스 힌트라고 보면,
`query_rewrite_select.c` 로 옮기는게 좋을 것 같아요.

3개 다 `query_rewrite_select.c` 로 옮기고, util.c 파일도 제거하는게 어떨까요?

sohee-dgist replied:
qo_is_oid_const, qo_rewrite_index_hints -> qo_rewrite_select.c
qo_rewrite_hidden_col_as_derived -> qo_rewrite_subquery.c

로 변경 하였습니다. 저 함수 자체가 subquery를 다시 쓰는 것이라..

youngjinj replied:
확인했습니다. 감사합니다.

Commit:
(refactor) 코드리뷰 수정사항 2차 반영
- query_rewrite_util.h 삭제
- query_rewrite_util.c 에 있는 함수들 대량 이동

shparkcubrid commented on src/optimizer/rewriter/query_rewrite_select.c:
```diff
@@ -0,0 +1,3267 @@
+
+
+static PT_NODE *qo_reset_location (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
+static PT_NODE *qo_optimize_queries (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
```
제거된 함수 입니다.

sohee-dgist replied:
변경 완료하였습니다.

Commit:
(refactor) rename된 함수 헤더에 남아 있는 것 삭제

- 주석 중에 rename 안 된 것 변경

youngjinj commented on src/optimizer/rewriter/query_rewrite.c:
```diff
@@ -0,0 +1,762 @@
+
+#include "query_rewrite.h"
+
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * query_rewrite_util.c - Query rewrite utils - Do Not Include Except This Folder
+ */
+
+#ident "$Id$"
+
+#include <assert.h>
+#include "query_rewrite.h"
```
중복입니다.

sohee-dgist replied:
확인했습니다.
rename refactoring이후 한번 더 커밋 올려 멘션 드리겠습니다.

sohee-dgist replied:
반영 완료하였습니다.

youngjinj replied:
3338564 commit까지 반영된 후에도 중복이 있습니다.

sohee-dgist replied:
로컬에서 오류가 있었던 것 같네요 수정했습니다

youngjinj replied:
확인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_select.c:
```diff
@@ -0,0 +1,3266 @@
+ *
+ */
+void
+qo_move_on_clause_of_explicit_join_to_where_clause (PARSER_CONTEXT * parser, PT_NODE ** fromp, PT_NODE ** wherep)
```
qo_analyze_path_join_pre와 qo_analyze_path_join 함수 사이에 위치하는 것이 좋지 않은 것 같습니다.

qo_analyze_path_join_pre 함수 위에 위치하거나
qo_analyze_path_join 함수 아래에 위치하게 하는 것은 어떤가요?

sohee-dgist replied:
저도 해당 의견이 맞아 보입니다. 수정하도록 하겠습니다.

youngjinj replied:
아직 위치는 변경되지 않은 것으로 확인했습니다.
이름이 변경된 `qo_move_on_of_explicit_join_to_where` 함수 위치 조정이 필요해보입니다.

sohee-dgist replied:
반영 완료 하였습니다.

youngjinj replied:
확인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite.h:
```diff
@@ -0,0 +1,174 @@
+#include "semantic_check.h"
+#include "view_transform.h"
+#include "virtual_object.h"
+#include "xasl_generation.h"
```
`pt_to_null_ordering` 함수 때문에 `xasl_generation.h` 파일의 include가 필요합니다.
rewriter에서 `xasl_generation.h` 파일을 필요로 한다는 것이 이상합니다.

`pt_to_null_ordering` 함수도 꼭 `xasl_generation.c` 파일에 정의 될 필요가 없어 보입니다.

`pt_to_null_ordering` 함수 선언을 `src/parser/parser.h` 로 옮기고,
정의를 `src/parser/parser_support.c` 로 옮기는 것은 어떤가요?

이렇게 되면 여기서 `xasl_generation.h` 파일을 include 할 필요가 없어집니다.

sohee-dgist replied:
저도 그렇게 생각합니다. 왜 PT Node 함수가 그곳에 달려있었는지 이해가 힘들군요. 반영하겠습니다.

sohee-dgist replied:
다시 보니 pt_to_함수들이 전부 xasl_generation.c에 위치해서 옮기기 애매한 것 같습니다.
include하더라도 사용하는 것이 맞는 듯 합니다.
다만 pt_to를 xasl_generation.c에 include해 놓고 header는 parser.h에 옮긴 케이스가 있습니다. 그런 케이스와 같이 수정 가능해 보입니다.

youngjinj replied:
전부 옮기자는 의견은 아니었습니다.
`query_rewrite.h`에서 `xasl_generation.h`를 include 하지 않기 위해서 `pt_to_null_ordering` 함수만 위치를 옮기자는 의견이었습니다.

`pt_to_null_ordering` 선언은 `parser.h`에 있지만 정의는 아직 `xasl_generation.c`에 있습니다.
`pt_to_null_ordering` 정의도 `src/parser/parser_support.c`로 옮기는 것은 어떤가요?

조금씩 `XASL_NODE`를 다루는 함수와 `PT_NODE`를 다루는 함수의 코드 분리가 필요하다고 생각합니다.

sohee-dgist replied:
pt_to 함수가 전부 xasl_generation 에 포함되어 있기 때문에 옮기기가 조금 조심스러워 집니다.
분리 하는 김에 전부 분리 하는 것은 어떨까요?

youngjinj replied:
리팩터링 범위가 query_rewriter.c 파일에 대한 것에서부터 시작했고, pt_to_null_ordering 함수의 변경은 그 파생 범위까지로 생각해볼 수 있는데, xasl_generation 관련 함수들을 옮겨오는 것은 범위가 너무 넓어지는 것 같습니다.

이 이슈를 마무리한 후에 의견을 들어서 단계적으로 진행하는게 좋겠다는 생각입니다.

sohee-dgist replied:
일단 옮겨 두었으나 위치가 애매하네용

youngjinj replied:
`src/parser/parser.h` 파일에서의 선언 위치나 `src/parser/parser_support.c` 파일에서의 정의 위치에 대해서는 판단하기 어렵습니다. 이 파일들의 변경이 필요할 때 판단할 수 있도록 남겨두어야 할 것 같습니다. 다른 의견이 없으시면 Resolve 해주시면 감사하겠습니다.

sohee-dgist replied:
넵 알겠습니다.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

Commit:
(rename) qo_do_auto_parameterize -> qo_auto_parameterize

Commit:
(refactor) query_rewrite_util.c 삭제 및 함수들 이동

Commit:
(refactor) 코드리뷰 수정사항 3차 반영

- renaming to short (gnu indent)
- clean header
- move pt_? functions to parser_tree_cl and parser_support

Commit:
(refactor) gnu indent 누락 체크

parser 파일 제외됨.

Commit:
(refactor) 헤더 정리

query_rewrite.h 헤더 정리

hgryoo committed:
[CBRD-25930] Fix cub_pl fails to restart on Windows (#5987)

http://jira.cubrid.org/browse/CBRD-25930

- Fix cub_pl would not start if the <db_name>_pl.info file, which stores the PID and port number, was not initialized before monitoring.
- Even after cubrid pl restart detected that cub_pl had failed to start and transitioned to the STOPPED state, m_failure_count was not reset. This caused an immediate transition to the FAILED_TO_INITIALIZE state.

beyondykk9 committed:
[CBRD-25951] PL/CSQL Static SQL Loop 내 CTE에서 SP 사용 시 Core Dump 발생 (#5988)

http://jira.cubrid.org/browse/CBRD-25951

youngjinj commented on src/optimizer/rewriter/query_rewrite.h:
```diff
@@ -0,0 +1,170 @@
+void qo_move_on_of_explicit_join_to_where (PARSER_CONTEXT * parser, PT_NODE ** fromp, PT_NODE ** wherep);
+void qo_rewrite_index_hints (PARSER_CONTEXT * parser, PT_NODE * statement);
+
+/* qo_auto_parameterize is defined in parser.h */
+void qo_do_auto_parameterize_limit_clause (PARSER_CONTEXT * parser, PT_NODE * node);
+void qo_do_auto_parameterize_keylimit_clause (PARSER_CONTEXT * parser, PT_NODE * node);
```
`qo_do_auto_parameterize` -> `qo_auto_parameterize` 로 함수명을 변경한 것과 관련해서
이 2개 함수명의 변경은 필요하지 않을까요?

`qo_do_auto_parameterize` 이름을 사용하는 주석이 2곳 있습니다.
같이 `qo_auto_parameterize`로 변경해야 할 것 같습니다.

sohee-dgist replied:
즉시 반영하도록 하겠습니다.

youngjinj replied:
홗인했습니다. 감사합니다.

youngjinj commented on src/optimizer/rewriter/query_rewrite_select.c:
```diff
@@ -0,0 +1,3661 @@
+ *   node(in):
+ */
+static int
+qo_is_oid_const (PT_NODE * node)
```
혹시 함수 선언은 어디에 있나요?

beyondykk9 committed:
[CBRD-25713] for merge : CHAR 타입 최대 길이 2048로 제한 (#5996)

http://jira.cubrid.org/browse/CBRD-25713

hyunikn committed:
[CBRD-25927] prepare once and close once over all loop iterations for other language constructs (#5949)

* loop optimization for various constructs using SQL
* not clearing loop optimizables at the semantic check of Body to apply the loop optimization to the SQL using statements and expressions in Block statements

YeunjunLee committed:
[CBRD-25956] Fix coredump when removing entry which does not exists (#5994)

http://jira.cubrid.org/browse/CBRD-25956

Fix coredump when removing entry which does not exists

mhoh3963 committed:
[CBRD-25945] fix memory leak of cas (#6000)

http://jira.cubrid.org/browse/CBRD-25945

- Fix to disconnect from db when calling signal handler

ctshim committed:
[CBRD-25902] Changed to use stx_alloc_struct() to support reuse of SP-related xasl (#5890)

http://jira.cubrid.org/browse/CBRD-25902

* Only changed the structure to support XASL reuse for SP functions as well

hgryoo committed:
[CBRD-25953] trivial: correct logic for clearing Context in PL server (#5989)

* add null checks for classLoaders in SessionClassLoaderGroup

* fix

* just clear, do not set null for classLoaders

* fix

* add synchronized for destroyContext ()

hgryoo committed:
[CBRD-25958] UDS를 사용하지 않고 랜덤 포트로 연결할 때 PL 재시작 시 PL 서버에 연결이 안되는 문제 수정 (#6006)

CBRD-25958

hgryoo committed:
[CBRD-25827] Fix segfault by dereferencing null (tfile_vfid_p->next) in `qmgr_free_list_temp_file` (#5975)

* [APIS-778]JDBC CUBRIDStatement.setFetchSize(int rows) throw wrong illegalArgumentException when rows value is zero

* check query entry in qmgr_clear_trans_wakeup(), null checks for qmgr_free_list_temp_file

* remove redundant newline

* pl's cursor should not be holdable

* revert

* m_query_entry is nullable if the query is not holdable

Commit:
(review) 코드리뷰 제안사항 4차 반영

ctshim committed:
[CBRD-25957] Memory leak problem in packer when using PL function (#6002)

http://jira.cubrid.org/browse/CBRD-25957

* fix memory leak to use PL functions
* When string compression is used for characters exceeding 255 characters.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

mhoh3963 committed:
[CBRD-25845] fix cause of coredump in debug mode (#6010)

http://jira.cubrid.org/browse/CBRD-25945

- call db_shutdown(), only if connected to db in the ux_database_shutdown().
- remove duplicate variable initialization code caused by ux_database_sutdown() being called in cas_free().

jongmin-won committed:
[CBRD-25939] Fix the issue where a synonym retains a duplicate name when the owner of the target class/vclass is changed. (#5935)

http://jira.cubrid.org/browse/CBRD-25939

When creating a Synonym, you cannot use the same unique name as an existing object (Class or Vclass).

However, there was an issue where changing the owner of a synonym's target class to match the synonym caused a duplicate unique name to be created.
This issue has now been fixed.

mhoh3963 committed:
[CBRD-25945] prevent additional execution during executing csession_end_session() (#6014)

hyunikn committed:
[CBRD-25929] loop opt for local routine (#6003)

* add another type of loop optimizable, LocalRoutineCall
* collect sql uses that are not contained in a loop but reachable by local routine calls from it
  init and fin the Statement before and after the main user code for the sql uses that are reachable from a loop body
* optimize SQL uses in the mutual recursive calls of local routines
* avoid getting connection for every recursive calls of Stored Procedures

hyunikn committed:
[CBRD-25659] add options to leave heap profile data on JVM OutOfMemory error (#6016)

* log max heap size and JVM options on PL server starts

* add two JVM options to profile heap memory on OutOfMemory error

* refinements
  . guarantee that closeSocket() is run by putting it in finally block
  . guarantee the the socket is closed when an exception occurs before ExecThread is run
  . put jvmheapdump.hprof in log directory

vimkim committed:
[CBRD-25966] strict_warnings_on/off.hpp - allow Compiler Warnings to Error, per file (#6013)

http://jira.cubrid.org/browse/CBRD-25966

* strict_warnings.h, with example perf.cpp

* fix mistake

* fix formatting

* make it intentionally fail

* fix newline issue (format)

* fix: sometimes not work in gcc

* fix new line at eof

* separate c/cpp

* rename to hpp

* fix h -> hpp

* feat!: add missing warnings

* now make build succeed

* recover perf.cpp

* fix: remove invalid and relocate c-only warnings

* enable error again to fix memory_wrapper ci false alarm

* rename: add _on

* docs: fix header file comment

* perf.cpp: fix typo in include

* check.yml: make strict_warnings_on/off.hpp pass memory-monitor-check

* fix comment indent

* add comments for github actions memory monitoring check

* remove perf.cpp change

* refactor: refer to the official documentation

* clean up warnings (c only, cpp only) with ifdef __cplusplus

hyunikn committed:
[CBRD-25659] slip: add a symbol to win/cubridsa/cubridsa.def (#6025)

minor: add a symbol to win/cubridsa/cubridsa.def

hgryoo committed:
[CBRD-25786] Add authenticate cache for procedures (#6011)

http://jira.cubrid.org/browse/CBRD-25786

Add a structure for caching authorization information for procedures, similar to AU_CLASS_CACHE, is introduced. This structure is used to retrieve the procedure's authorization information directly.

- Procedure cache: A map of type <MOP, vector*> is used and managed in the same way as AU_CLASS_CACHE.
- Modify authenticate_cache::update to be applicable to both tables and procedures.
- Add a get_procedure_cache_bits() function to allocate and return the procedure's cache bits structure.
- Add a procedure cache removal routine to authenticate_cache::flush.
- Modify au_check_procedure_authorization() to use the procedure's authorization cache instead of querying authorization information via check_execute_authorization_by_query().

hgryoo committed:
[CBRD-25968] Put in request cubmethod::header into bootstrap_request struct (#6029)

http://jira.cubrid.org/browse/CBRD-25968

There is no need to create a header every time a bootstrap_request is sent to the PL server; instead, it should be included in the bootstrap_request struct. Additionally, a Valgrind report indicated that std::move(header) might have caused uninitialized bytes due to the object's lifetime not being guaranteed. This could be a memory wrapper issue, but since the reason is unclear, defensive code will be added.

- Include cubmethod::header as a member variable in bootstrap_request.
- Implement the constructor and pack function.

ctshim committed:
[CBRD-25964] fix CHECK_OVER_CODE_VALUE is always true (#6018)

http://jira.cubrid.org/browse/CBRD-25964

* Check and fix compile warning messages

ctshim committed:
[CBRD-25964] clear CHECK_OVER_CODE_VALUE macro (#6036)

http://jira.cubrid.org/browse/CBRD-25964

*  clear  CHECK_OVER_CODE_VALUE macro

hornetmj committed:
[CBRD-25970] print errcode in the server error log for debugging purpose (#6032)

http://jira.cubrid.org/browse/CBRD-25970

mhoh3963 committed:
[CBRD-25945] fix coredump (#6030)

http://jira.cubrid.org/browse/CBRD-25945

- modify db_end_session() as the following
  * if session_id is DB_EMPTY_SESSION, then it returns immediately.
  * if keep_session is false and csession_end_session() is success, session_id is set to DB_EMPTY_SESSION.
- There is a place where the ref_count is reduced without locking session_p in another function of the server, so there may be a case where the ref_count is 0 in the session_state_destroy() function. So the assert is removed and a defensive code is added.
- modify to check if tdes is NULL in the  logtb_set_current_user_active()

beyondykk9 committed:
[CBRD-25922] DBLink Auto-commit OFF Handling (#5919)

http://jira.cubrid.org/browse/CBRD-25922

airnet73 committed:
[CBRD-25737] Add user schema to SELECT NEXT_VALUE statement (#6037)

http://jira.cubrid.org/browse/CBRD-25737

Create serial with a general account and perform unloaddb with a dba account. Add user schema missing from select statement.

shparkcubrid committed:
[CBRD-25947] Problem with changing rownum to orderby_num even though order_by is deleted during view merge (#6038)

In the process of converting order_by during view merge, rownum and orderby_num are changed unnecessarily. Conversion is performed only for the following cases.

case 1: order by of mainquery is newly added ==> rownum of mainquery is changed to orderby_num
case 2: order by of subquery is totally removed ==> orderby_num of subquery is changed to rownum

hyunikn committed:
[CBRD-25138] report the first syntax error when parsing PL/CSQL programs (#6040)

* throw SyntaxError on the first syntax error
  . it was on the last one because of the error recovery mechanism of ANTLR

mhoh3963 committed:
[CBRD-25945] Modify to clear only client memory without requesting the server when a signal occurs. (#6042)

http://jira.cubrid.org/browse/CBRD-25945

- add db_shutdown_without_request_to_server() for clearing only client memory without requesting the server.
- add definition for window os

hgryoo committed:
[CBRD-25881] Modify the SP loading process not to have the same Class in /java and /java_static (#6031)

http://jira.cubrid.org/browse/CBRD-25881

During loadjava, check both the /java and /java_static directories. If a class with the same name already exists, ensure overwrite checking is performed.

- When loading a Java SP, modify the logic to first attempt to load the class from the ServerClassLoader, where the JNI anchor class might be stored. If not found, fall back to the ContextClassLoader (getOldClassLoader).
- Move the following code to the beginning of session-based class loading (PL/CSQL, Java SP)
- Modify the loadjava logic to:
  - Check both /java and /java_static directories for classes with the same name.
  - Remove the existing class if it is being overwritten.

mhoh3963 committed:
[CBRD-25932] fix a bug in sql where the prompt did not change on disconnecting status at window os (#6044)

http://jira.cubrid.org/browse/CBRD-25932

beyondykk9 committed:
[CBRD-25980] Prepared Statement error for TIME values ​​starting with '13:' (#6048)

http://jira.cubrid.org/browse/CBRD-25980

jongmin-won committed:
[CBRD-25909] (Spec Change) Transfer permissions granted by the previous owner to the new owner when changing the owner of a Class/Vclass or Stored Procedure. (#6033)

http://jira.cubrid.org/browse/CBRD-25909

Duplicate conditions in the update_auth_for_new_owner function will be consolidated for optimization.

beyondykk9 committed:
[CBRD-25358] rev (no error for system class): error handling when using for update clause for inline view (#6047)

http://jira.cubrid.org/browse/CBRD-25358

hgryoo committed:
[CBRD-25732] Create pl_session if it is empty when setting PL_CONTEXT system parameter (#6051)

http://jira.cubrid.org/browse/CBRD-25732

When setting the PL_CONTEXT system parameter, if a pl_session has not been created, the parameter change is ignored. Modify the behavior to create a pl_session if it does not exist

hgryoo committed:
[CBRD-25817] Ensure PL session parameters, including DBMS_OUTPUT, persist across interrupts and server restarts (#6022)

http://jira.cubrid.org/browse/CBRD-25817

Modify the system so that PL-related session parameters, including DBMS_OUTPUT, persist in the following scenarios:
- Interrupts
- PL server restart

*DBMS_OUTPUT.enable(), disable()*

- Add the `METHOD_CALLBACK_SET_PL_SESSION_PARAM` command.
- Send the `METHOD_CALLBACK_SET_PL_SESSION_PARAM` command when enabling or disabling `DBMS_OUTPUT`.
- Update `DBMS_OUTPUT` session parameters in `pl_session` via `callback_set_pl_session_param()`.

*Modify Relevant Protocols and Base Code*

- Define parameters used exclusively in the PL session independently from the DB server session parameters: Add `enum class sys_param_id`.
- Modify the constructor and value assignment logic for the `sys_param` object in `pl_session.cpp`, and add unpacking logic.
- (Java) Add packing logic for the `SysParam` object.

*Determine PL Server Restart or Interrupts*

- Use the `check_reloading_pl_context_required()` function to detect PL server restart or interrupt events.
  1. On an interrupt, set `m_last_conn_epoch` to `-1`.
  2. Detect PL server shutdown or restart by monitoring `connection_pool` epoch changes:
     - `m_last_conn_epoch != conn->get_epoch()`: The session's last successful connection does not match the latest epoch.
     - `!conn->is_valid()`: There was no previous connection in this session, but the newly assigned connection does not match the latest epoch.
  3. Manage connection success history in the session via `m_last_conn_epoch` in `claim_connection()` and `release_connection()`.

- Using the check in **2-2**, determine whether the PL server was restarted while using a previously established or newly acquired connection from the pool.
- Using **2-3**, ensure session parameters are correctly updated in the PL server's session object (`Context`).
- Use `obtain_session_parameters()` to generate session parameters to be sent during PL calls.

hyunikn committed:
[CBRD-25982] %FOUND and %NOTFOUND cursor attributes must be null between open and fetch (#6046)

%FOUND and %NOTFOUND cursor attributes must be null between open and fetch operations

hyunikn committed:
[CBRD-25175] remove line and column from PL/CSQL compile error messages (#6054)

remove line and column from PL/CSQL compile error messages

mhoh3963 committed:
[CBRD-25992] revise cubrid.sh(csh) (#6056)

http://jira.cubrid.org/browse/CBRD-25992

- remove unused environment variables (SHLIB_PATH, LIBPATH)
- unify environment variable declaration in cubrid.sh to "export environment variable=setting value"
- modify setup.sh and CPack.STGZ_Header.sh.in due to changes in cubrid.sh
- add comment for tunning glibc memory library

hgryoo committed:
[CBRD-25748] Add trace stats for SP (#6049)

http://jira.cubrid.org/browse/CBRD-25748

In TRACE, when an SP call is present, the collected data now includes SP call execution time, fetch, ioread, and calls.

- Add the following statistics to `func_stat`:
  - `time`: Evaluation time
  - `fetches`: Number of fetches
  - `ioreads`: Number of I/O reads
  - `calls`: Number of SP calls

- Add new performance statistics to `perf_monitor`:
  - `PSTAT_REGU_NUM_CALL_EVALS`: Number of SP calls
  - `PSTAT_REGU_EVAL_TIME_10USEC`: Total evaluation time (in 10µs units)
  - `PSTAT_REGU_NUM_FETCHES`: Number of fetches during evaluation
  - `PSTAT_REGU_NUM_IOREADS`: Number of I/O reads during evaluation

- Collect SP call statistics in:
  - `fetch` operations
  - `query_executor`

- Print call statistics in `query_dump` only when `calls > 0`

airnet73 committed:
[CBRD-25503] Allows connection to a Broker that has not been set up. (#6050)

http://jira.cubrid.org/browse/CBRD-25503

- change parameter name from ACCESS_CONTROL_BEHAVIOR_FOR_EMPTYBROKER to ACCESS_CONTROL_DEFAULT_POLICY.
- change the location of the setting value from broker section to the common section.

hgryoo committed:
[CBRD-25983] Modify native lib loading for JNI to be allowed only for SP that use a statically loaded class loader (#6057)

http://jira.cubrid.org/browse/CBRD-25983

In the class loader that dynamically loads classes, if a native library for JNI is loaded, a duplicate native library loading may occur on subsequent calls, requiring a restart of the PL server. Therefore, the class loader for dynamic loading should be modified to prevent it from loading native libraries.

- Modify to throw an SecurityException in SpSecurityManager.checkLink() when dynamic class loader try to load native library.

mhoh3963 committed:
[CBRD-25986] fix memory leak (#6058)

http://jira.cubrid.org/browse/CBRD-25986

kangmin5505 committed:
[CBRD-25979] au_change_class_owner_including_partitions 함수에서 db_private_strdup 함수 실패 시, AU_ENABLE 누락

Fixed missing AU_ENABLE call on db_private_strdup failture in
au_change_class_owner_including_partitions

beyondykk9 committed:
[CBRD-25985] memory leak at cubscan::method::scanner::init() (#6061)

http://jira.cubrid.org/browse/CBRD-25985

beyondykk9 committed:
[CBRD-25998] core dump at perfmon_server_dump_stats() (#6067)

http://jira.cubrid.org/browse/CBRD-25998

beyondykk9 committed:
[CBRD-25922] rev: DBLink Auto-commit OFF Handling (#6064)

http://jira.cubrid.org/browse/CBRD-25922

beyondykk9 committed:
[CBRD-25996] max_query_per_tran=1 does not apply in PL/CSQL SP. (#6063)

http://jira.cubrid.org/browse/CBRD-25996

hyunikn committed:
[CBRD-25981] bugfix: applied Java string escape twice. fixed to apply it just once. (#6065)

bugfix: applied Java string escape twice. fixed to apply it just once.

hyunikn committed:
[CBRD-25659] parametrize JVM heap dump on OOM (#6060)

* adding a system parameter to turn on/off JVM heap dump on OOM

sohee-dgist committed:
[CBRD-25500] Fix Invalid XASL Error in ORDERBY_NUM() Query (#5952)

http://jira.cubrid.org/browse/CBRD-25500

Include host variable count in XASL hash text

hgryoo committed:
[CBRD-25163] Term Change: Procedure Language → Procedural Language (#6074)

http://jira.cubrid.org/browse/CBRD-25163

Trivial: Replace all "Procedure Language" -> "Procedural Language".

hyunikn committed:
[CBRD-26000] fix the error message on serial value query failures (#6077)

fix the error message on serial value query failures

shparkcubrid committed:
[CBRD-25990] The problem where for orderby_num() in inline views gets removed due to view merging (#6075)

http://jira.cubrid.org/browse/CBRD-25990

1. Add copy logic for orderby_for in mq_update_order_by().
2. If there is an aggregate function or distinct in the main query, exclude adding hidden columns (as-is: skip the entire routine)

jongmin-won committed:
[CBRD-25834] Fix security issue exposing procedure source code to EXECUTE-only users (#6071)

http://jira.cubrid.org/browse/CBRD-25834

Fixes a security issue where the code column in the db_stored_procedure view was exposed to users with only EXECUTE privilege, potentially leaking the source code of stored procedures.

ctshim committed:
[CBRD-26002] In system parameter.c, PRM_ID_JAVA_STORED_PROCEDURE_JVM_OPTIONS is incorrectly written as PRM_ID_STORED_PROCEDURE_JVM_OPTIONS (#6084)

http://jira.cubrid.org/browse/CBRD-26002

* typo:  PRM_ID_JAVA_STORED_PROCEDURE_JVM_OPTIONS

beyondykk9 committed:
[CBRD-26001] [Regression][shell] Core dumped in dblink_end_tran (#6082)

http://jira.cubrid.org/browse/CBRD-26001

shparkcubrid committed:
[CBRD-25990] The problem where for orderby_num() in inline views gets removed due to view merging (#6089)

http://jira.cubrid.org/browse/CBRD-25990

If both the main query and the subquery have an order by clause, view merge is possible, but if the subquery has orderby_for, view merge is not possible. Modify the assert.

hornetmj committed:
[CBRD-25068] temp removal cub_alloc wrapper for QA test purpose (#6095)

http://jira.cubrid.org/browse/CBRD-25068

This commit will be reverted after QA checks the momory leak issues.

H2SU committed:
[CBRD-25970] Revert "print errcode in the server error log for debugging purpose" (#6096)

Revert "[CBRD-25970] print errcode in the server error log for debugging purpose (#6096)"

This reverts commit 3504a3702c39c36f941e0bfcdbe404fcde072a9a.

hwany7seo committed:
[APIS-1034] Update JDBC version from 11.3.0 to 11.3.1. (#6086)

http://jira.cubrid.org/browse/APIS-1034

- SubModule Update (Update JDBC version to 11.3.1)

hornetmj committed:
Revert "[CBRD-25068] temp removal cub_alloc wrapper for QA test purpose" (#6098)

Revert "[CBRD-25068] temp removal cub_alloc wrapper for QA test purpose (#6095)"

This reverts commit 7722090f31ec3a4aa0dfad5d2430cb1909b6cfd6.

beyondykk9 committed:
[CBRD-25358] rev (error for system class / view in inline view): error handling when using for update clause for system tables such as db_root, dual, etc. (#6094)

http://jira.cubrid.org/browse/CBRD-25358

sohee-dgist committed:
[CBRD-25500] Change Bind Var Count when Caching Subquery (#6097)

http://jira.cubrid.org/browse/CBRD-25500

changed bind_variable counting logic when using subquery cache

mhoh3963 committed:
[CBRD-26008] add fflush(stdout) after printing version on debug mode (#6101)

http://jira.cubrid.org/browse/CBRD-26010

mhoh3963 committed:
Revert "[CBRD-26008] add fflush(stdout) after printing version on debug mode" (#6105)

This reverts commit 128f2112badc1580f01b672d99ba11bf05810701.

mhoh3963 committed:
[CBRD-26008] add fflush(stdout) after printing version on debug mode (#6106)

http://jira.cubrid.org/browse/CBRD-26008

beyondykk9 committed:
[CBRD-25358] rev (error check at resolve names): error handling when using for update clause for inline view (#6107)

http://jira.cubrid.org/browse/CBRD-25358

hyunikn committed:
[CBRD-26014] bugfix: give a legal identifier name to a column when necessary (#6109)

* bugfix: give a legal identifier name to a column in a select list whose name cannot be a PL/CSQL Id and hence cannot be referenced

mhoh3963 committed:
[CBRD-26016] optional clear objects at the signal handler of cas (#6111)

http://jira.cubrid.org/browse/CBRD-26016

- Since signal handlers occur randomly, it is possible for the task's object clearing and the signal handler's object clearing to occur at the same time.
- Therefore, I modify that objects that can be cleared during operation are excluded to clearing it in the signal handler.

InChiJun committed:
[CBRD-26010] Fix backup hang with different max_clients settings in [common] and [@db] sections in conf file (#6102)

http://jira.cubrid.org/browse/CBRD-26010

hgryoo committed:
[CBRD-26019] Remove incorrect PARSER_VARCHAR *val declarations in plcsql_text_part (#6116)

http://jira.cubrid.org/browse/CBRD-26019

trivial:
Since pt_append_string() returns a char*, declaring PARSER_VARCHAR *val was misleading and unnecessary.
This PR revises from PARSER_VARCHAR* to char* for the val.

airnet73 committed:
[CBRD-26025] Error occurs when sending NULL value to DBLink via JDBC (#6118)

http://jira.cubrid.org/browse/CBRD-26025

H2SU committed:
[CBRD-26015] Fix flashback utility unable to restart when previous execution isn't terminated after forced quit (Ctrl+C) (#6110)

http://jira.cubrid.org/browse/CBRD-26015

YeunjunLee committed:
[CBRD-26018] Fix incorrect database name used during Server Revive restart (#6114)

http://jira.cubrid.org/browse/CBRD-26018

Fix incorrect database name used during Server Revive restart

hgryoo committed:
[CBRD-26021] Fix NULL default when using non-pseudo column expressions in a stored function' parameter (#6115)

http://jira.cubrid.org/browse/CBRD-26021

- Add evaluation routine for non-pseudo column expression to create a new PT_VALUE node from the evauluated DB_VALUE
- Add 'LOG' function to SymbolStack

tw-kang committed:
[CBRD-26030] upgrade action runner image version refer to the deprecated image. (#6121)

http://jira.cubrid.org/browse/CBRD-26030

* modify ubuntu version from 20.04 to 24.04 refer to deprecated image.
* cppcheck package upgrade 2.13.0

kisoo-han committed:
[CBRD-26020] [CMS] fix error in wait-for-lock, gettransactioninfo (#6127)

fix error in string_tokenize () for wait-for-lock field in gettransactioninfo (), when a multiple transaction IDs are listed

vimkim committed:
[CBRD-26026] perf: Reduce sizeof DB_VALUE by 8 bytes (#6119)

http://jira.cubrid.org/browse/CBRD-26026

fix!(dbtype_def): swap int and char* to reduce 8 bytes in DB_CHAR, as a result reduce DB_VALUE by 8 bytes

Optimize struct layout: reduce DB_VALUE size from 72 to 64 bytes

Reordered members in the DB_CHAR union by moving `int` before `char *`, 
reducing DB_CHAR size from 40 to 32 bytes due to better alignment.

This change also reduces the size of:
- DB_ENUM_ELEMENT: 48 -> 40 bytes
- DB_DATA (union): 48 -> 40 bytes
- DB_VALUE: 72 -> 64 bytes

kisoo-han committed:
[CBRD-26028] remove minus in lob locator suffix (#6128)

http://jira.cubrid.org/browse/CBRD-26028

- remove unexpected minus character in lob locator

hgryoo committed:
[CBRD-25985] Fix memory leak caused in method_scan (#6120)

http://jira.cubrid.org/browse/CBRD-25985

- replace C++ containers with C POD structs to fix memory leak.

H2SU committed:
[CBRD-26005] Add log record type EOL check (#6129)

Add an explicit EOL check in the log reader to stop advancing past end‑of‑log and prevent attempts to fetch non‑existent pages.

vimkim committed:
[CBRD-26026] Revert - perf: Reduce sizeof DB_VALUE by 8 bytes (#6132)

http://jira.cubrid.org/browse/CBRD-26026

Revert "[CBRD-26026] perf: Reduce sizeof DB_VALUE by 8 bytes (#6119)"

This reverts commit 7e233340d93f280f05f348930610f77905d87ef0.

InChiJun committed:
[CBRD-26027] Separate worker pool for backup read tasks (#6131)

http://jira.cubrid.org/browse/CBRD-26027

airnet73 committed:
[CBRD-26031] When entering Korean characters using bind variables in DBLINK, Korean characters are broken (#6137)

http://jira.cubrid.org/browse/CBRD-26031

Purpose
gateway와 이기종 DB 간에는 odbc unicode driver를 이용한다. 이기종 DB에 한글을 insert 할때 한글을 unicode로 변환하여 전송해야한다. 그런데 bind 변수를 사용하는 경우, unicode로 변환을 하고 있지 않아 한글이 깨지는 현상이 발생함

Implementation
bind 변수 값을 unicode로 변환하여 전송하도록 변경함

hgryoo committed:
[CBRD-26034] Fix memory leak in session_set_pl_session () (#6136)

http://jira.cubrid.org/browse/CBRD-26034

Delete session->pl_session_p regardless of any thread type in session_stop_attached_threads().

hgryoo committed:
[CBRD-26034] Do not delete pl_session when conn is just down, CAS would reconnect the session again (#6143)

http://jira.cubrid.org/browse/CBRD-26034

InChiJun committed:
[CBRD-26027] [regression] Ensure the minimum number of cores and workers, and prevent unlocking an uninitialized mutex object (#6145)

http://jira.cubrid.org/browse/CBRD-26027

hyunikn committed:
[CBRD-26040] do not push for dblink query (#6148)

do not push for dblink query

hyunikn committed:
[CBRD-26042] prohibit DBLink in PL/CSQL Static SQL DML (#6149)

* prohibit DBLink in PL/CSQL Static SQL DML

mhoh3963 committed:
[CBRD-26048] fix error when connecting to slave db through cas in HA environment (#6158)

http://jira.cubrid.org/browse/CBRD-26048

mhoh3963 committed:
[CBRD-26049] preparing for guava project (#6162)

http://jira.cubrid.org/browse/CBRD-26049

- change version : 11.5.0
- change codeowner

tw-kang committed:
[CBRD-26051] Remove build-windows workflow on circleci (#6160)

http://jira.cubrid.org/browse/CBRD-26051

Purpose
Remove the build-windows workflow so that Windows builds are not supported from version guava(11.5).

mhoh3963 committed:
[CBRD-25870] support au_disable_passwords () for all library (#6178)

http://jira.cubrid.org/browse/CBRD-25870

tw-kang committed:
[CUBRIDQA-1263] add comment trigger for circleci trigger on PR (#6182)

http://jira.cubrid.org/browse/CUBRIDQA-1245
http://jira.cubrid.org/browse/CUBRIDQA-1263

add comment trigger for circleci trigger on PR
Usage : /run [all | sql | plcsql | medium | shell] [baseline]

tw-kang committed:
[CUBRIDQA-1263] fix bug for branch name (#6183)

http://jira.cubrid.org/browse/CUBRIDQA-1263

* fix bug for branch name

tw-kang committed:
[CUBRIDQA-1245] add test_shell on circleci. (#6035)

http://jira.cubrid.org/browse/CUBRIDQA-1245

Usage : /run [all | sql | medium | plcsql | shell ] [baseline]

The required job(SQL, medium) must be executed before the merge.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

kangmin5505 committed:
[CBRD-26061] Ignore #error directives during cppcheck in GitHub Actions (#6176)

http://jira.cubrid.org/browse/CBRD-26061

* suppress preprocessorErrorDirective in cppcheck

mhoh3963 committed:
[CBRD-25992] fix an error where two CUBRID environment variables were created in cubrid.sh (#6186)

http://jira.cubrid.org/browse/CBRD-25992

hornetmj committed:
[CBRD-25602] Fix assert() when executing drop table with vacuum_disable=y (#5637)

http://jira.cubrid.org/browse/CBRD-25602

hornetmj committed:
[CBRD-25645] [bug-fix] Assert when restarting server after committing some XA transactions (#5677)

http://jira.cubrid.org/browse/CBRD-25645

InChiJun committed:
[CBRD-25664] Excluded -1 from the valid range of the page_flush_interval system parameter (#5638)

http://jira.cubrid.org/browse/CBRD-25664

InChiJun committed:
[CBRD-25696] Remove unused variables in lock_manager.c (#5670)

http://jira.cubrid.org/browse/CBRD-25696

InChiJun committed:
[CBRD-25652] Fix the incorrect calculation of log_Gl.hdr.avg_ntrans when transactions with top operations exist during checkpoint execution (#5606)

http://jira.cubrid.org/browse/CBRD-25652

InChiJun committed:
[CBRD-25541] Fixed unnecessary heap operations in show access status by passing NULL to the class_oid parameter (#5613)

http://jira.cubrid.org/browse/CBRD-25541

InChiJun committed:
[CBRD-25510] Removed unnecessary loop in heap_get_insert_location_with_lock() function execution (#5654)

http://jira.cubrid.org/browse/CBRD-25510

InChiJun committed:
[CBRD-25028] Disable heap_chnguess_realloc() function, which is inaccessible from heap_Guesschn (#5688)

http://jira.cubrid.org/browse/CBRD-25028

InChiJun committed:
[CBRD-25734] Fix assert on CUBRID server stop with incomplete XA transactions (#6085)

http://jira.cubrid.org/browse/CBRD-25734

vimkim committed:
[CBRD-25784] Support Clang Compiler (2nd trial) (#6189)

http://jira.cubrid.org/browse/CBRD-25784

Supports Clang Compiler

Fix a few fatal errors in modern compilers (that used to be allowed to compile in old compilers like gcc8).

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter

tw-kang committed:
[CUBRIDQA-1274] fix bug for test split (#6188)

http://jira.cubrid.org/browse/CUBRIDQA-1274


* revert globing pattern to legacy

InChiJun committed:
[CBRD-25510] Revert - Removed unnecessary loop in heap_get_insert_location_with_lock() function execution (#6199)

This reverts commit 64b1b6f491cd2adabb83c9e52cc1f772be7ed0dc.

xmilex-git committed:
[CBRD-25447] Support parallel heap scan (#6141)

Purpose
To improve the performance of heap scans, we plan to add a parallel heap scan feature to CUBRID.

Implementation
This is a merge pull request for the feature branch.

Remarks
Since this is a merge pull request for the feature branch, the code has already been reviewed.
It includes the following pull requests.
#5850
#6093
#6091
#6154
#6166
#6155

youngjinj committed:
[CBRD-25991] XASL Generation Failure for Sort-Merge and Hash Joins When Functional Index Expressions Are Not Join Predicates (#6142)

http://jira.cubrid.org/browse/CBRD-25991

Fix XASL generation failure for Sort-Merge and Hash joins when functional index expressions are used in the WHERE clause and not as join predicates.

ctshim committed:
[CBRD-26017] clear warnings: -Wfromat, -Wsing-compare, -Wrestrict,-Wwrite-strings (#6113)

http://jira.cubrid.org/browse/CBRD-26017

* clear compile warnings:
   -Wfromat
   -Wsing-compare
   -Wrestrict
   -Wwrite-strings
   -Wformat-truncation=


* [CBRD-26017] clear warnings: -Wfromat, -Wsing-compare, -Wrestrict,-Wwrite-strings

* include

* build

* build

* build

* src/

* build

* build

* build

* test

* PRI?64

* SCNu64

* for cppcheck

* for cppcheck

ctshim committed:
[CBRD-26007] clear -Wreorder warning (#6099)

http://jira.cubrid.org/browse/CBRD-26007

[CBRD-26007] clear Wreorder warning message
* clear -Wreorder warning

xmilex-git committed:
[CBRD-26087] [Regression] Core dumped in qfile_reopen_list_as_append_mode at src/query/list_file.c:1348 (#6204)

Purpose

A core dump occurs in debug mode due to assert(0) when the list-id is not initialized inside the reopen() function. This change addresses the issue.

Implementation

Modified the logic to prevent reopen from being called when list_id is uninitialized or when the operation is not based on the mergable-list method.

Remarks

N/A

ctshim committed:
[CBRD-26059] Remove comments about server information from the rewritten string of a DBLINK query (#6194)

http://jira.cubrid.org/browse/CBRD-26059

* [CBRD-26059] Remove comments on server information from the rewited string of a DBLINK query
   - Hide dblink connection information when using SERVER name in query
   - Do not show connection information when obtaining rewritten query string such as query plan.
   - Since the information was used when querying XASL cache, it is not exposed in plan dump either, so readability is eliminated by using hash.

ctshim committed:
[CBRD-26055] Remove code and files related to esql tool (#6198)

http://jira.cubrid.org/browse/CBRD-26055

* [CBRD-26055] Remove code and files related to esql tool
  - Remove esql related tools and source files

sohee-dgist committed:
[CBRD-25741] Converting uppercase SQL identifiers to lowercase (#6180)

http://jira.cubrid.org/browse/CBRD-25741

Converted uppercase SQL identifiers to lowercase

kangmin5505 committed:
[CBRD-25709] Infinite Loop in DEFERRED Trigger with Recursive Structure (Maximum Trigger Depth Not Enforced)

JIRA: http://jira.cubrid.org/browse/CBRD-25709

Purpose
Fix an issue where recursive structures in DEFERRED triggers could lead to infinite loops due to the Maximum Trigger Depth limit not being enforced.

Implementation
Applied a Breadth-First Search (BFS)-like approach to group trigger executions by layer, treating each layer as a single depth level. This aligns the behavior of DEFERRED triggers with that of BEFORE and AFTER triggers, ensuring an error is raised when the maximum trigger depth is exceeded.

beyondykk9 committed:
[CBRD-26033] Processing UPDATE DBLink queries in trigger actions (#6164)

http://jira.cubrid.org/browse/CBRD-26033

H2SU committed:
[CBRD-24462] Improve the logic that determines the number of threads to be used when running the backupdb utility, and additionally print the number of threads specified by administrator (#5537)

http://jira.cubrid.org/browse/CBRD-24462

tw-kang committed:
[CUBRIDQA-1268] use cache for build and test jobs (#6203)

http://jira.cubrid.org/browse/CUBRIDQA-1268

use compiler cache 'ccache'
use testcase cache.

HyunukLee committed:
[CBRD-26058] Fix error in CASE expression with logical type (#6170)

http://jira.cubrid.org/browse/CBRD-26058

Add handling for logical-type operators with non-logical inputs in CASE expressions

HyunukLee committed:
[CBRD-25975] Handle unmatched special characters in argument-requiring SQL hints (#6139)

http://jira.cubrid.org/browse/CBRD-25975

Handle parsing failures in SQL hints when special characters like ` or " are used incorrectly in arguments.

youngjinj committed:
[CBRD-25846] Support PARTITION HASH JOIN (#6012)

http://jira.cubrid.org/browse/CBRD-25846

### Cost Calculation
- File hash joins have the same cost as other hash join methods, without additional I/O cost.

### XASL Generation
- Refactored code related to XASL generation for hash joins.
   - Renamed `gen_outer_hash_join` to `gen_hashjoin`.
   - Moved `PROJECTION_INFO`-related logic into separate functions.
   - Moved `QFILE_LIST_MERGE_INFO`-related logic into a separate function.

### Executor
1. Refactored hash join execution into a separate file and restructured it to be context-based.
2. If file hashing is required in a single (non-partitioned) hash join, the input is partitioned so that each partition can use a hybrid hash join if possible.
3. The number of partitions is determined by the side with fewer tuples between the outer and inner inputs.

### SQL Trace
- SQL trace output separately shows partitioning and hash join execution.

### Parameters
- When `max_hash_list_scan_size` differs from the default, its value is included in the `sql_hash_text`, which affects XASL cache reuse.

### Miscellaneous
- The following predefined macros can be set to 1 in each source file to enable additional testing, debugging, or profiling features, without modifying `CMakeLists.txt` to add `add_definitions()`.

   **src/optimizer/query_planner.c**
   - `TEST_DUMP_PLAN_COST`
   - `TEST_HASH_JOIN_ENABLE`
   - `TEST_HASH_JOIN_FORCE_ENABLE`

   **src/query/query_hash_join.h**
   - `HASHJOIN_PROFILE_TIME`
   - `HASHJOIN_DUMP_PARTITION`
   - `HASHJOIN_DUMP_HASH_TABLE`
   - `HASHJOIN_DUMP_BUILD`
   - `HASHJOIN_DUMP_PROBE`

   **src/query/query_hash_scan.h**
   - `HASH_LIST_SCAN_DUMP_HASH_TABLE`
   - `HASH_LIST_SCAN_DUMP_FILE_HASH`

Hamkua committed:
[CBRD-25759] Enhance Partition Table (#6173)

Purpose
Enhances features related to partition tables.

Implementation
Includes changes that have already been reviewed and merged into the feature/partition-table branch.

The following pull requests are included:
#5398
#5689
#5682
#6112

xmilex-git committed:
[CBRD-26088] [Regression][shell] Core dumped in file_tempcache_pop_tran_file at src/storage/file_manager.c:9521 (#6212)

In the original CUBRID source, during the execution of a holdable query, if the value of qentry_p->list_id->page_cnt was 0, the query was considered to have not created a temp file, and thus the corresponding temp file was not preserved.

However, with the introduction of parallel heap scan, a temp file containing results may exist even if it is not directly linked to the list_id. Such files also need to be managed along with the corresponding qentry.

Accordingly, the logic that skipped preserving the temp file when qentry_p->list_id->page_cnt was 0 has been removed.

xmilex-git committed:
[CBRD-26090] [Regression][shell] Core dumped in src/query/parallel/px_heap_scan/px_heap_scan_manager.cpp:392 (#6213)

http://jira.cubrid.org/browse/CBRD-26090

Purpose
This update addresses the following six issues:

Failure in scan_open_parallel_heap_scan() when using file_get_num_user_pages()
An assert() failure occurs in debug mode if the query is interrupted during the call to scan_open_parallel_heap_scan(), which uses file_get_num_user_pages().
Memory deallocation mismatch with memory_mapper
An assert() failure occurs when attempting to deallocate memory allocated with malloc() using db_private_free() during the deletion of a memory_mapper.
assert() failure due to NULL tfile_vfid in open_list()
If an interrupt occurs at the moment open_list() attempts to acquire tfile_vfid, a write operation may still be attempted despite tfile_vfid being NULL, resulting in an assert() failure.
assert() failure due to concurrent access to mvcc_snapshot in child threads
When a child thread calls scan_start_scan(), it may try to access the mvcc_snapshot concurrently, leading to an assert() failure.
Loss of tempcache entry due to concurrent access
When a tempcache entry is removed from the free list and pushed to the transaction list, multiple threads may access it simultaneously, resulting in the loss of the tempcache entry.
Incorrect status check for child thread termination
The main thread misinterprets the child thread’s termination status as scan_ended instead of task_ended, proceeding without proper memory cleanup. This can cause the query to end while the child thread is still active, resulting in an assert() failure in restrack().

ctshim committed:
[CBRD-25978] refactoring to system parameters (#6090)

http://jira.cubrid.org/browse/CBRD-25978

* [CBRD-25978] refactoring system_parameter.c
    - Use PARAM_ID instead of PARAM_NAME
    - Add prm_Def_session_idx to access cached session parameters at once without loop search

ctshim committed:
[CBRD-26038] Results are different when using broker_log_converter/broker_log_runner for VARBIT type (#6150)

http://jira.cubrid.org/browse/CBRD-26038

* [CBRD-26038] Results are different when using broker_log_converter/broker_log_runner for VARBIT type
  - Resolve the problem of truncating the last byte in SQL logs for varbit and bit types
  - Fix the error in handling varbit types in broker_log_conveter and broker_log_runner

xmilex-git committed:
[CBRD-26090] [Regression][shell] Core dumped in src/query/parallel/px_heap_scan/px_heap_scan_manager.cpp:392 (#6225)

http://jira.cubrid.org/browse/CBRD-26090

Purpose

In previous PR (#6213), we added the code that make access to freed memory.
So we modified it.

InChiJun committed:
[CBRD-25767] Implement specifying multiple classes for heap dump with diagdb -d 9 (#6191)

http://jira.cubrid.org/browse/CBRD-25767

kangmin5505 committed:
[CBRD-26097] memory leak of au_change_class_owner function (#6223)

Purpose:
Fixed a memory leak caused by an early return after memory allocation for `sub_partitions` 
in `au_change_class_owner`. The leak occurs when returning before reaching the cleanup section.

Implementation:
Replaced early return with `goto end` to ensure `sub_partitions` is properly freed before exit.

sohee-dgist committed:
Merge branch 'CUBRID:develop' into RND-2404_refactor_rewriter
